src/App.tsx

```diff
@@ -1,24 +1,218 @@
+
+    try {
+      await updateTodo(id, isChecked);
+
+      setTodos(todos.map(todo => {
+        if (todo.id === id) {
+          return { ...todo, completed: isChecked };
+        }
+
+        return todo;
+      }));
+
+      if (isChecked) {
+        setRemainingTodos(remainingTodos - 1);
+      } else {
+        setRemainingTodos(remainingTodos + 1);
+      }
+    } catch (error) {
+      setErrorMessage('Unable to update a todo');
+    } finally {
+      setUpdatingIds([]);
+    }
+  };
+
+  const handleToggleAllTodos = async () => {
+    const todoIds = todos.map(todo => todo.id);
+
+    setUpdatingIds([...updatingIds, ...todoIds]);
+
+    try {
+      await Promise.all(todos.map(
+        todo => updateTodo(todo.id, !areAllTodosCompleted),
+      ));
+
+      const newTodos = todos.map(todo => (
+        { ...todo, completed: !areAllTodosCompleted }
+      ));
+
+      setTodos(newTodos);
+      setAreAllTodosCompleted(!areAllTodosCompleted);
+      setRemainingTodos(newTodos.filter(todo => !todo.completed).length);
+    } catch (error) {
+      setErrorMessage('Unable to update todos');
+    } finally {
+      setUpdatingIds([]);
+    }
+  };
+
+  const handleClearCompleted = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+    const todoIds = completedTodos.map(todo => todo.id);
+
+    setUpdatingIds([...updatingIds, ...todoIds]);
+
+    try {
+      await Promise.all(completedTodos.map(todo => deleteTodo(todo.id)));
+
+      setTodos(todos.filter(todo => !todo.completed));
+    } catch (error) {
+      setErrorMessage('Unable to delete a completed todos');
+    } finally {
+      setUpdatingIds([]);
+    }
+  };
+
+  const handleAddTodo = async (title: string) => {
+    if (title.trim().length === 0) {
+      setErrorMessage('Title can\'t be empty');
+
+      return;
+    }
+
+    const newTempTodo: Todo = {
+      id: 0,
+      title,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo(newTempTodo);
+    setIsCreatingTodo(true);
+
+    try {
+      const newTodo: Todo = await createTodo(title, USER_ID);
+
+      setTodos([...todos, newTodo]);
+      setRemainingTodos(remainingTodos + 1);
+    } catch (error) {
+      setErrorMessage('Unable to add a todo');
+    } finally {
+      setIsCreatingTodo(false);
+      setTempTodo(null);
+    }
+  };
+
+  const handleRemoveTodo = async (id: number, completed: boolean) => {
+    setUpdatingIds([...updatingIds, id]);
+
+    try {
+      await deleteTodo(id);
+
+      setTodos(todos.filter(todo => todo.id !== id));
+      if (!completed) {
+        setRemainingTodos(remainingTodos - 1);
+      }
+    } catch (error) {
+      setErrorMessage('Unable to delete a todo');
+    } finally {
+      setUpdatingIds([]);
+    }
+  };
+
+  const handleDismissNotification = () => {
+    setErrorMessage('');
+  };
+
+  const filteredTodos = todos.filter(todo => {
+    switch (filter) {
+      case 'all':
+        return true;
+      case 'active':
```

Comment:
let's create an enum for that 
```suggestion
      case FilterType.ACTIVE:
```